File: Legacy_CSharp_Projects/TarsCli.Tests/Commands/AutoImprovementCommandTests.cs
using System;
using System.CommandLine;
using System.CommandLine.Invocation;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Moq;
using TarsEngine.Metascripts;
using TarsCli.Commands;
using Xunit;

namespace TarsCli.Tests.Commands
{
    public class AutoImprovementCommandTests
    {
        [Fact]
        public void Constructor_ShouldCreateCommand()
        {
            // Arrange
            var loggerMock = new Mock<ILogger<AutoImprovementCommand>>();
            var executorMock = new Mock<IMetascriptExecutor>();
            
            // Act
            var command = new AutoImprovementCommand(loggerMock.Object, executorMock.Object);
            
            // Assert
            Assert.Equal("auto-improve", command.Name);
            Assert.NotNull(command.Handler);
        }
        
        [Fact]
        public void Constructor_NullLogger_ShouldThrowArgumentNullException()
        {
            // Arrange
            var executorMock = new Mock<IMetascriptExecutor>();
            
            // Act & Assert
            Assert.Throws<ArgumentNullException>(() => new AutoImprovementCommand(null, executorMock.Object));
        }
        
        [Fact]
        public void Constructor_NullExecutor_ShouldThrowArgumentNullException()
        {
            // Arrange
            var loggerMock = new Mock<ILogger<AutoImprovementCommand>>();
            
            // Act & Assert
            Assert.Throws<ArgumentNullException>(() => new AutoImprovementCommand(loggerMock.Object, null));
        }
        
        [Theory]
        [InlineData("all", "TarsCli/Metascripts/Improvements/auto_improvement_pipeline.tars")]
        [InlineData("code-quality", "TarsCli/Metascripts/Improvements/code_quality_analyzer.tars")]
        [InlineData("documentation", "TarsCli/Metascripts/Improvements/documentation_generator.tars")]
        [InlineData("tests", "TarsCli/Metascripts/Improvements/test_generator.tars")]
        public async Task RunAutoImprovement_ValidTarget_ShouldExecuteCorrectMetascript(string target, string expectedMetascriptPath)
        {
            // Arrange
            var loggerMock = new Mock<ILogger<AutoImprovementCommand>>();
            var executorMock = new Mock<IMetascriptExecutor>();
            
            executorMock
                .Setup(e => e.ExecuteMetascriptAsync(It.IsAny<string>(), It.IsAny<object>()))
                .ReturnsAsync(new MetascriptExecutionResult { Success = true });
            
            var command = new AutoImprovementCommand(loggerMock.Object, executorMock.Object);
            
            // Create a mock file system to make File.Exists return true
            var fileSystemMock = new Mock<IFileSystem>();
            fileSystemMock.Setup(fs => fs.FileExists(It.IsAny<string>())).Returns(true);
            
            // Act
            await InvokeCommandHandler(command, target, false, false);
            
            // Assert
            executorMock.Verify(e => e.ExecuteMetascriptAsync(
                It.Is<string>(s => s == expectedMetascriptPath),
                It.IsAny<object>()),
                Times.Once);
        }
        
        [Fact]
        public async Task RunAutoImprovement_InvalidTarget_ShouldLogError()
        {
            // Arrange
            var loggerMock = new Mock<ILogger<AutoImprovementCommand>>();
            var executorMock = new Mock<IMetascriptExecutor>();
            
            var command = new AutoImprovementCommand(loggerMock.Object, executorMock.Object);
            
            // Act
            await InvokeCommandHandler(command, "invalid-target", false, false);
            
            // Assert
            loggerMock.Verify(
                x => x.Log(
                    LogLevel.Error,
                    It.IsAny<EventId>(),
                    It.Is<It.IsAnyType>((v, t) => v.ToString().Contains("Unknown target: invalid-target")),
                    It.IsAny<Exception>(),
                    It.IsAny<Func<It.IsAnyType, Exception, string>>()),
                Times.Once);
            
            executorMock.Verify(e => e.ExecuteMetascriptAsync(
                It.IsAny<string>(),
                It.IsAny<object>()),
                Times.Never);
        }
        
        private async Task InvokeCommandHandler(Command command, string target, bool dryRun, bool verbose)
        {
            // Get the handler
            var handler = command.Handler as CommandHandler;
            
            // Create a context
            var context = new InvocationContext(null);
            
            // Set the values in the context
            context.BindingContext.SetValueForOption(command.Options[0], target);
            context.BindingContext.SetValueForOption(command.Options[1], dryRun);
            context.BindingContext.SetValueForOption(command.Options[2], verbose);
            
            // Invoke the handler
            await handler.InvokeAsync(context);
        }
    }
    
    // Interface for mocking File.Exists
    public interface IFileSystem
    {
        bool FileExists(string path);
    }
}

File: v1/parked_legacy/TarsEngine.FSharp.Cli/TarsEngine.FSharp.Cli/Commands/Tests/AutoImprovementCommandTests.fs
namespace TarsEngine.FSharp.Cli.Commands.Tests

open System
open System.IO
open System.Threading.Tasks
open Microsoft.Extensions.Logging
open TarsEngine.FSharp.Core.Metascript
open TarsEngine.FSharp.Cli.Commands
open Xunit
open Moq

/// <summary>
/// Tests for the AutoImprovementCommand.
/// </summary>
module AutoImprovementCommandTests =
    /// <summary>
    /// Tests that the command executes successfully.
    /// </summary>
    [<Fact>]
    let ``AutoImprovementCommand executes successfully`` () =
        // Arrange
        let loggerMock = Mock<ILogger<AutoImprovementCommand>>()
        let metascriptExecutorMock = Mock<IMetascriptExecutor>()
        
        // Setup the metascript executor to return a successful result
        metascriptExecutorMock.Setup(fun m -> m.ExecuteMetascriptAsync(It.IsAny<string>(), It.IsAny<obj>()))
            .Returns(Task.FromResult({ Success = true; ErrorMessage = null }))
        
        // Create the command
        let command = AutoImprovementCommand(loggerMock.Object, metascriptExecutorMock.Object)
        
        // Create the options
        let options = CommandOptions.createDefault()
        
        // Act
        let result = command.ExecuteAsync(options).Result
        
        // Assert
        Assert.True(result.Success)
        Assert.Equal(0, result.ExitCode)
        Assert.Contains("Auto-improvement pipeline completed successfully", result.Message)
    
    /// <summary>
    /// Tests that the command handles errors.
    /// </summary>
    [<Fact>]
    let ``AutoImprovementCommand handles errors`` () =
        // Arrange
        let loggerMock = Mock<ILogger<AutoImprovementCommand>>()
        let metascriptExecutorMock = Mock<IMetascriptExecutor>()
        
        // Setup the metascript executor to return a failed result
        metascriptExecutorMock.Setup(fun m -> m.ExecuteMetascriptAsync(It.IsAny<string>(), It.IsAny<obj>()))
            .Returns(Task.FromResult({ Success = false; ErrorMessage = "Test error" }))
        
        // Create the command
        let command = AutoImprovementCommand(loggerMock.Object, metascriptExecutorMock.Object)
        
        // Create the options
        let options = CommandOptions.createDefault()
        
        // Act
        let result = command.ExecuteAsync(options).Result
        
        // Assert
        Assert.False(result.Success)
        Assert.NotEqual(0, result.ExitCode)
        Assert.Contains("Auto-improvement pipeline failed", result.Message)
        Assert.Contains("Test error", result.Message)
    
    /// <summary>
    /// Tests that the command handles different targets.
    /// </summary>
    [<Fact>]
    let ``AutoImprovementCommand handles different targets`` () =
        // Arrange
        let loggerMock = Mock<ILogger<AutoImprovementCommand>>()
        let metascriptExecutorMock = Mock<IMetascriptExecutor>()
        
        // Setup the metascript executor to return a successful result
        metascriptExecutorMock.Setup(fun m -> m.ExecuteMetascriptAsync(It.IsAny<string>(), It.IsAny<obj>()))
            .Returns(Task.FromResult({ Success = true; ErrorMessage = null }))
        
        // Create the command
        let command = AutoImprovementCommand(loggerMock.Object, metascriptExecutorMock.Object)
        
        // Create the options with a specific target
        let options = CommandOptions.createDefault()
        let options = CommandOptions.withOptions (Map.ofList [("target", "code-quality")]) options
        
        // Act
        let result = command.ExecuteAsync(options).Result
        
        // Assert
        Assert.True(result.Success)
        Assert.Equal(0, result.ExitCode)
        Assert.Contains("Auto-improvement pipeline completed successfully", result.Message)
        
        // Verify that the correct metascript was executed
        metascriptExecutorMock.Verify(fun m -> m.ExecuteMetascriptAsync("TarsCli/Metascripts/Improvements/code_quality_analyzer.tars", It.IsAny<obj>()), Times.Once())
